pkg/operator/controller/canary/controller.go

```diff
@@ -0,0 +1,231 @@
+		APIVersion: "apps/v1",
+		Kind:       "Deployment",
+		Name:       deployment.Name,
+		UID:        deployment.UID,
+		Controller: &trueVar,
+	}
+
+	haveService, service, err := r.ensureCanaryService(deploymentRef)
+	if err != nil {
+		errors = append(errors, err)
+		return result, utilerrors.NewAggregate(errors)
+	}
+	if !haveService {
+		errors = append(errors, fmt.Errorf("failed to get canary service"))
+		return result, utilerrors.NewAggregate(errors)
+	}
+
+	if haveRoute, _, err := r.ensureCanaryRoute(service); err != nil {
+		if err != nil {
+			errors = append(errors, err)
+			return result, utilerrors.NewAggregate(errors)
+		}
+		if !haveRoute {
+			errors = append(errors, fmt.Errorf("failed to get canary route"))
+			return result, utilerrors.NewAggregate(errors)
+		}
+	}
+
+	return result, nil
+}
+
+// enqueueRequestForIngressCanary returns reconcile requests for
+// properly labeled ingress canary resources in the given namespace.
+func enqueueRequestForIngressCanary(namespace string) handler.EventHandler {
+	return &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
+			labels := a.Meta.GetLabels()
+			if _, ok := labels[manifests.OwningIngressCanaryCheckLabel]; ok {
+				log.Info("queueing ingress canary", "related", a.Meta.GetSelfLink())
+				return []reconcile.Request{
+					{
+						NamespacedName: types.NamespacedName{
+							Namespace: namespace,
+							Name:      a.Meta.GetName(),
+						},
+					},
+				}
+			} else {
+				return []reconcile.Request{}
+			}
+		}),
+	}
+}
+
+// enqueueRequestForDefaultIngressController returns canary controller
+// reconcile requests for the default ingress controller.
+func enqueueRequestForDefaultIngressController(namespace string) handler.EventHandler {
+	return &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(a handler.MapObject) []reconcile.Request {
+			if cmp.Equal(a.Meta.GetName(), manifests.DefaultIngressControllerName) {
+				log.Info("queueing ingress canary", "related", a.Meta.GetSelfLink())
+				return []reconcile.Request{
+					{
+						NamespacedName: types.NamespacedName{
+							Namespace: namespace,
+							Name:      manifests.DefaultIngressControllerName,
+						},
+					},
+				}
+			} else {
+				return []reconcile.Request{}
+			}
+		}),
+	}
+}
+
+// Config holds all the things necessary for the controller to run.
+type Config struct {
+	Namespace   string
+	CanaryImage string
+}
+
+// reconciler handles the actual canary reconciliation logic in response to
+// events.
+type reconciler struct {
+	Config
+
+	client client.Client
+	cache  cache.Cache
+}
+
+// Switch the current RoutePort that the route points to.
+// Use this function to periodically update the canary route endpoint
+// to verify if the router has wedged.
+func (r *reconciler) rotateRouteEndpoint(service *corev1.Service, current *routev1.Route) (*routev1.Route, error) {
```

Comment:
Looks like nothing calls this right now.  Is that going to be in phase 2?

Reply:
yes, `rotateRouteEndpoints` and `chooseRandomServicePort` are for phase 2.

Reply:
Why not remove the unused funcs and store them in a WIP phase 2 branch instead of including them in this PR?

Reply:
In retrospect I probably should not have committed functions that aren't called by anything.

That being said, the functions have already been reviewed, and will be utilized soon.
@danehans would it be OK if we merge them now with the understanding that they will be used as a part of NE-199 phase 2, which is coming shortly?

Reply:
I prefer that they're removed, but I won't block the PR if they stay. I suggest at least adding a TODO to any unimplemented functions in case phase 2 gets pushed out for some reason.